api/bucketsd/service.go

```diff
@@ -2069,16 +2078,17 @@ func (s *Service) PushPaths(server pb.APIService_PushPathsServer) error {
 				if err != nil {
 					return saveWithErr(fmt.Errorf("adding bucket link: %v", err))
 				}
-				ctx2, err := s.updateOrAddPin(ctx, path.New(buck.Path), dir)
+				newCtx, err = s.updateOrAddPin(newCtx, path.New(buck.Path), dir)
 				if err != nil {
 					return saveWithErr(fmt.Errorf("updating bucket pin: %v", err))
 				}
-				ctxLock.Lock()
-				ctx = ctx2
-				ctxLock.Unlock()
 			}
 			buck.Path = dir.String()
 			buck.UpdatedAt = time.Now().UnixNano()
+			buck.SetMetadataAtPath(res.path, tdb.Metadata{
+				UpdatedAt: buck.UpdatedAt,
+			})
+			buck.UnsetMetadataWithPrefix(res.path + "/")
```

Comment:
Since processing of `res` is handled serially, we can safely alter the bucket right after the UnixFS DAG is altered.